File: CS/CustomSummaryHelper.cs
using System;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraGrid.Views.Grid;

namespace WeightedAverages {
	public class CustomSummaryHelper {

		public static object GetWeightedAverage(GridView view, string weightField, string valueField) {
			const decimal undefined = 0;
			if(view == null) return undefined;
			GridColumn weightCol = view.Columns[weightField];
			GridColumn valueCol = view.Columns[valueField];
			if(weightCol == null || valueCol == null) return undefined;
			
			try {
				decimal totalWeight = 0, totalValue = 0;
				for(int i = 0; i < view.DataRowCount; i++) {
					if(view.IsNewItemRow(i)) continue;
					object temp;
					decimal weight, val;
					temp = view.GetRowCellValue(i, weightCol);
					weight = (temp == DBNull.Value || temp == null) ? 0 : Convert.ToDecimal(temp);
					temp = view.GetRowCellValue(i, valueCol);
					val = (temp == DBNull.Value || temp == null) ? 0 : Convert.ToDecimal(temp);

					totalWeight += weight;
					totalValue += weight * val;
				}
				if(totalWeight == 0) return undefined;
				return totalValue / totalWeight;
			}
			catch {
				return undefined;
			}
		}		
	}
}

File: VB/CustomSummaryHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraGrid.Views.Grid

Namespace WeightedAverages
	Public Class CustomSummaryHelper

		Public Shared Function GetWeightedAverage(ByVal view As GridView, ByVal weightField As String, ByVal valueField As String) As Object
			Const undefined As Decimal = 0
			If view Is Nothing Then
				Return undefined
			End If
			Dim weightCol As GridColumn = view.Columns(weightField)
			Dim valueCol As GridColumn = view.Columns(valueField)
			If weightCol Is Nothing OrElse valueCol Is Nothing Then
				Return undefined
			End If

			Try
				Dim totalWeight As Decimal = 0, totalValue As Decimal = 0
				For i As Integer = 0 To view.DataRowCount - 1
					If view.IsNewItemRow(i) Then
						Continue For
					End If
					Dim temp As Object
					Dim weight, val As Decimal
					temp = view.GetRowCellValue(i, weightCol)
					If (temp Is DBNull.Value OrElse temp Is Nothing) Then
						weight = 0
					Else
						weight = Convert.ToDecimal(temp)
					End If
					temp = view.GetRowCellValue(i, valueCol)
					If (temp Is DBNull.Value OrElse temp Is Nothing) Then
						val = 0
					Else
						val = Convert.ToDecimal(temp)
					End If

					totalWeight += weight
					totalValue += weight * val
				Next i
				If totalWeight = 0 Then
					Return undefined
				End If
				Return totalValue / totalWeight
			Catch
				Return undefined
			End Try
		End Function
	End Class
End Namespace
